torch/_inductor/lowering.py

```diff
@@ -3146,6 +3150,38 @@ def load_source_val():
     return view(result_flat, self.get_size())
 
 
+@register_lowering(aten._unsafe_masked_index, type_promotion_kind=None)
+def _unsafe_masked_index(self, mask, indices, fill):
+    ranges, _unsafe_index_fn = index_impl_helper(self, indices, check=False)
+    mask_loader = mask.make_loader()
+
+    def inner_fn(idx):
+        return ops.masked(mask_loader(idx), lambda: _unsafe_index_fn(idx), fill)
+
+    return Pointwise.create(
+        device=self.get_device(),
+        dtype=self.get_dtype(),
+        inner_fn=inner_fn,
+        ranges=ranges,
+    )
+
+
+@register_lowering(aten._unsafe_masked_index_put)
+def _unsafe_masked_index_put(x, mask, indices, values, accumulate=False):
+    masked_value = where(mask, values, 0)
+    shape = x.get_size()
+    clamped_indices = [
+        clamp(indices[i], -shape[i], shape[i] - 1) if indices[i] else None
+        for i in range(len(indices))
+    ]
```

Comment:
We can also do this via a masked operation since tl.atomic_add [accepts a mask](https://triton-lang.org/main/python-api/generated/triton.language.atomic_add.html#triton.language.atomic_add), but I don't know if inductor supports it as-is.
Can you see if it would work, and if it does, implement it with a masked index_add. Otherwise, just leave a comment and we'll look into it in a follow-up PR.